Add a useResponseClassTransformer global option

epiphone opened the pull request:
Hi, first of all thanks for a great library!

By default response values are coerced to plain objects with [class-transformer](https://github.com/pleerock/class-transformer). When returning large objects (#226) or values with complex serialization logic (e.g. Mongoose documents #149) you might opt for the default `toJSON` handler instead.

This PR adds a global `useResponseClassTransformer` option parameter for toggling response transformation. It defaults to `true`, and is overridden by `classTransformer: false`.

Also related to #179.

19majkel94 commented on src/driver/BaseDriver.ts:
```diff
@@ -129,14 +135,14 @@ export abstract class BaseDriver {
     protected transformResult(result: any, action: ActionMetadata, options: Action): any {
         // check if we need to transform result
-        const shouldTransform = (this.useClassTransformer && result != null) // transform only if enabled and value exist
+        const shouldTransform = (this.useClassTransformer && this.useResponseClassTransformer && result != null) // transform only if enabled and value exist
```
not updated comment

19majkel94 commented:
Isn't that the feature that @NoNameProvided promised to implement a few months ago? 😉

NoNameProvided commented:
Yes, it was. Shame I haven't had the time. I haven't reviewed this either because I think we had agreed on a slightly different approach, but I am not sure, and I don't have time to read it back now. I have a lot of thing on my table now, so I struggle to find any time to spend on my hobbies (including this lib) right now.

epiphone commented:
Hi, I updated the PR to better match with @NoNameProvided's solution [outlined here](https://github.com/typestack/routing-controllers/issues/226#issuecomment-317025358), i.e. added `transformRequest` and `transformResponse` options to controller and route decorators. Gotta agree this approach seems more future-proof :smile: 

What do you think?

19majkel94 commented on src/decorator-options/ControllerOptions.ts:
```diff
@@ -0,0 +1,8 @@
+/**
+ * Extra options that apply to each controller action.
+ */
+export interface ControllerOptions extends TransformationOptions {
```
I think that you should implement `TransformationOptions`, not extending it - it would be ready for future extension/changes because you can't extend multiple interfaces.

19majkel94 commented on test/functional/action-options.spec.ts:
```diff
@@ -0,0 +1,70 @@
+        assertRequest([3001, 3002], "post", "default", { firstName: "Umed", lastName: "Khudoiberdiev" }, response => {
+            expect(initializedUser).to.be.instanceOf(User);
+            expect(response).to.have.status(200);
+            expect(response.body.lastName).to.be.defined;
```
You've set `transformResponse: false` but you expect `lastName` to be defined even if `lastName is excluded when class-transformer is disabled`... I don't get it 😕

19majkel94 commented on test/functional/controller-options.spec.ts:
```diff
@@ -0,0 +1,88 @@
+        }
+
+        @JsonController("/transform", {transformRequest: true, transformResponse: true})
+        class NoResponseTransformController {
```
`NoResponseTransformController`? **No?** 😆

19majkel94 commented on test/functional/controller-options.spec.ts:
```diff
@@ -0,0 +1,88 @@
+        assertRequest([3001, 3002], "post", "default", { firstName: "Umed", lastName: "Khudoiberdiev" }, response => {
+            expect(initializedUser).to.be.instanceOf(User);
+            expect(response).to.have.status(200);
+            expect(response.body.lastName).to.be.defined;
```
Again:
- default: `...lastName).to.be.defined;`
- enable: `...lastName).not.to.be.undefined;`
What's going on? Isn't transform enabled by default?

19majkel94 commented on test/functional/action-options.spec.ts:
```diff
@@ -0,0 +1,70 @@
+        assertRequest([3001, 3002], "post", "override", { firstName: "Umed", lastName: "Khudoiberdiev" }, response => {
+            expect(initializedUser).not.to.be.instanceOf(User);
+            expect(response).to.have.status(200);
+            expect(response.body.lastName).to.be.defined;
```
Shouldn't you use `.exist` here? I can't see `defined` in chai documentation 😕

epiphone commented:
Thanks for the feedback @19majkel94, lots of good points in there :+1: . I refactored the tests, let me know if there's anything else you'd like to see changed.

I did have some trouble with class-transformer's global metadata storage: calling `defaultMetadataStorage.clear()` ([here](https://github.com/typestack/routing-controllers/blob/master/test/functional/class-transformer-options.ts#L32) for example) deletes the class-transformer metadata of all defined classes, so I had to move the `User` class definitions inside a `before`-block and [assign it to a variable](https://github.com/typestack/routing-controllers/pull/329/commits/c4089663728183e965f78ba226abddf3171192c1#diff-d0a8cf7e0ad5016427a3046b660399fcR36) in the outer scope. The end result is still reasonably clean, I think, but let me know if you'd have a better solution.

19majkel94 commented:
Yeah, it clears because you call `clear` 😆 But I don't see the need to do this - pleerock does only:
```ts
 before(() => {
            getMetadataArgsStorage().reset();
```
In class-transformer-options tests. So you should only once declare the model and clear only the routing-controllers metadata.

epiphone commented:
Because the way it was set up, after a test module finished it would not only clear the metadata of its own test classes, but also of all the classes declared in the following tests. For example after `class-transformer-options.ts` finished, it wiped out `User`'s metadata in `global-options.spec.ts` which would break the test. That's why I ended up delaying the class declarations until the `before` block; I'd be happy to refactor it though if you have an alternative solution in mind.

19majkel94 commented:
So the problem was that pleerock uses the class-transformer only in one test so clearing metadata wasn't affect other tests but now you have tests also in other files so it clears the metadata of all tests, right? 😄

epiphone commented:
Exactly :+1: . I took a peek at class-transformer tests, seems like they're handling things in a similar way, declaring classes in the upper `describe` and `it` blocks.

19majkel94 commented:
Ok, you have `user.lastName || "default";` in your tests but I can't see a test case that is using it. Please make a one more test case that have `transformRequest: true` and `transformResponse: false` to see that it's working and we are ready to merge 😉

epiphone commented:
Hey, sorry for the delay, got kinda busy over the holidays! I added the test case and removed `lastName = "default";` where it's unused.

19majkel94 commented:
I'm not sure if @pleerock will like the options object inside decorator so I will wait for his review before merging this PR 😉

vekexasia commented:
Hello, i'm jumping here to understand if this was holded back due to an alternative way to use both JSON controllers and "Raw" controllers that returns buffers

epiphone commented:
I'm also curious about what's holding this back. Care to comment @NoNameProvided ?

f4z3k4s commented:
@pleerock, @NoNameProvided could you please have a look at this? This is still an existing issue :(

kunokdev commented:
Is there anything happening regarding this?

CSLTech commented:
Is there still something holding this back @NoNameProvided @pleerock ?

MohammedSiddiqui commented:
What's the update on this ? :sweat_smile:

hallsamuel90 commented:
Also curious about the status of this?

jotamorais commented:
@hallsamuel90, @MohammedSiddiqui @epiphone, @CSLTech, @kunokdev, @f4z3k4s, @vekexasia, @MichalLytek.

I'm planning to merge this PR and release this tomorrow night among some other fixes:

- #543 fix(Interceptors): order of global interceptors 
- #518 Input-validation bypass vulnerability 
- #438 class-validator arbitrary bypass vulnerability
- #536 Add support for wildcard "all" routes
- #546 chore: add another example for using the response directly
- #329 Add a useResponseClassTransformer global option
- #226 Performance Issue with big json result
- #149 Problem with mongoose model serialization 
- #179 use classTransformer only for input
- #301 Controller inheritance
- #147 Controller Inheritance support

yurkagon commented:
Still waiting :)
